### ISSUE TITLE
Failed to get console log 

### ISSUE THREAD

**Issue by Francesko90** [ContainX/openstack4j#795]

**Comment by Francesko90** [comment]
Follow up

**Comment by Francesko90** [comment]
Possible connection https://github.com/ContainX/openstack4j/issues/839

**Comment by lorenzo-biava** [comment]
This still happens with the `openstack4j-resteasy` connector, v3.0.3.

It looks like Resteasy client is looking for a `MessageBodyWriter` for the `String` type (the console log API is one of the few that is implemented with a JSON String instead of a POJO).
However, the only two [registered providers](https://github.com/ContainX/openstack4j/blob/3.0.3/connectors/resteasy/src/main/java/org/openstack4j/connectors/resteasy/ResteasyClientFactory.java#L29) are `JacksonJsonProvider` and `InputStreamProvider`.
The latter is obviously skipped and the `JacksonJsonProvider` is also unable to write a `String` body (see the check being made by the provider - we are currently using [v2.3.2](https://github.com/FasterXML/jackson-jaxrs-providers/blob/jackson-jaxrs-providers-2.3.2/base/src/main/java/com/fasterxml/jackson/jaxrs/base/ProviderBase.java#L539), but should be the same for the version [v2.7.3](https://github.com/FasterXML/jackson-jaxrs-providers/blob/jackson-jaxrs-providers-2.7.3/base/src/main/java/com/fasterxml/jackson/jaxrs/base/ProviderBase.java#L535) defined in the connector's POM).

I'm currently not able to test it using a newer version of the `JacksonJsonProvider`, but looking at the code  it seems like the behavior is the same.

@auhlig Could you let me know if you think it is just a problem with the Jackson library version or it is a problem to be fixed in the client configuration (say adding a `MessageBodyWriter` for the JSON String type in the `ResteasyClientFactory`)? Thanks.

**Comment by auhlig** [comment]
Hi @lorenzo-biava,
Many thanks for investigating.
I hoped it's just a problem with the jackson lib. that would be resolved for us by upgrading.
As you describe it that might not be the case. 
How much work would it be to fix it in os4j?

**Comment by lorenzo-biava** [comment]
Well I think it would be quite fast, using one of two options:
1) As far as i can see, the JSON String way (i.e. the `BaseOpenStackService.Invocation#json` method) is only used by few APIs. It might be removed and replaced with standard entity encoding (which I personally think would be much more consistent and readable).
Indeed, I temporarily solved the console log problem defining a proper POJO:
```java
@JsonRootName("os-getConsoleOutput")
class ConsoleOutputParams implements ModelEntity {

	private static final long serialVersionUID = 1L;

	public ConsoleLogParams(long length) {
	 this.length=length;
	}
	
	@JsonProperty("length")
	private long length;
	
	public long getLength() {
		return length;
	}	
}
```
and then using the `entity` method instead of `json`: `ConsoleOutput c = post(ConsoleOutput.class, uri("/servers/%s/action", serverId)).entity(new ConsoleOutputParams(numLines)).execute();`.

It looks like there are 3 other APIs that would require such refactoring: 
![image](https://cloud.githubusercontent.com/assets/5684713/23297772/14a12af8-fa7b-11e6-9f63-fc325282d453.png)

2) Another way, which will keep the JSON String way, would be to define an ad-hoc `MessageBodyWriter` for RestEasy and register it to the client.

Something like this:
```java
@Provider
@Produces({ "application/*+json", "text/json" })
public class JsonStringBodyWriter implements MessageBodyWriter<String> {

	@Override
	public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
		// We just need to write a JSON String, nothing else
		return type == String.class;
	}

	@Override
	public long getSize(String value, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
		// deprecated by JAX-RS 2.0
		return 0;
	}

	@Override
	public void writeTo(String jsonString, Class<?> type, Type genericType, Annotation[] annotations,
			MediaType mediaType, MultivaluedMap<String, Object> httpHeaders, OutputStream entityStream)
			throws IOException, WebApplicationException {

		DataOutputStream dos = new DataOutputStream(entityStream);
		dos.writeUTF(jsonString);
	}
}
```

`ResteasyClientFactory:`
```java
public ResteasyClientFactory() {
        super();
        
        addContextResolver(new CustomContextResolver());
        
        jsonProvider = new JacksonJsonProvider();
        addMessageBodyReader(jsonProvider);
        addMessageBodyWriter(jsonProvider);

        addMessageBodyWriter(new JsonStringBodyWriter());
```

I haven't tested this way, but, except for some improvements, it should be working.

**Comment by auhlig** [comment]
I agree, the 1st option looks simple and readable. Would prefer that. 
Would there be any reason, we should take the more complex 2nd approach?

What do you think @vinodborole, @gondor?

**Comment by vinodborole** [comment]
I agree with 1st option; it looks clean

**Comment by lorenzo-biava** [comment]
If you're going with the 1st option, I'd like to mention that the length parameter in the [Nova API](https://developer.openstack.org/api-ref/compute/?expanded=show-console-output-os-getconsoleoutput-action-detail#show-console-output-os-getconsoleoutput-action) is actually optional (and if not specified, will return the full log). Since this might be a useful feature (I'm using it, for instance), I would remove the automatic length setting in the `getConsoleOutput` implementation and change the `ModelEntity ` to allow for optional length:
```java
@JsonRootName("os-getConsoleOutput")
class ConsoleOutputParams implements ModelEntity {

	private static final long serialVersionUID = 1L;

	@JsonProperty("length")
	private Long length;

	/**
	 * To get the full console output.
	 */
	public ConsoleOutputParams() {
		this.length = null;
	}

	/**
	 * To get the last <code>length</code> lines of the console output.
	 * @param length the number of lines to retrieve.
	 */
	public ConsoleOutputParams(Long length) {
		this.length = length;
	}

	public Long getLength() {
		return length;
	}

}
```

**Comment by auhlig** [comment]
Very nice @lorenzo-biava!
Since you already did the work, would you like to submit a PR for that?

**Comment by lorenzo-biava** [comment]
Ok, I can put it in a PR, but I will limit to the ConsoleOutput. You may want to verify that the other APIs that use the `json` method have the same problem and fix it separately...

**Comment by lorenzo-biava** [comment]
@auhlig May I ask you when will this be released? I was hoping to see it in v3.0.4 quite soon...

**Comment by auhlig** [comment]
Good point. Long time since last release. Will do the needful :)
